components/style/values/specified/box.rs

```diff
@@ -7,15 +7,429 @@
+            Display::RubyText |
+            Display::RubyBaseContainer |
+            Display::RubyTextContainer
+        )
+    }
+
+    /// Returns whether this "display" value is a ruby level container.
+    #[cfg(feature = "gecko")]
+    pub fn is_ruby_level_container(&self) -> bool {
+        matches!(*self,
+            Display::RubyBaseContainer |
+            Display::RubyTextContainer
+        )
+    }
+
+    /// Convert this display into an equivalent block display.
+    ///
+    /// Also used for style adjustments.
+    pub fn equivalent_block_display(&self, _is_root_element: bool) -> Self {
+        match *self {
+            // Values that have a corresponding block-outside version.
+            Display::InlineTable => Display::Table,
+            Display::InlineFlex => Display::Flex,
+
+            #[cfg(feature = "gecko")]
+            Display::InlineGrid => Display::Grid,
+            #[cfg(feature = "gecko")]
+            Display::WebkitInlineBox => Display::WebkitBox,
+
+            // Special handling for contents and list-item on the root
+            // element for Gecko.
+            #[cfg(feature = "gecko")]
+            Display::Contents | Display::ListItem if _is_root_element => Display::Block,
+
+            // These are not changed by blockification.
+            Display::None |
+            Display::Block |
+            Display::Flex |
+            Display::ListItem |
+            Display::Table => *self,
+
+            #[cfg(feature = "gecko")]
+            Display::Contents |
+            Display::FlowRoot |
+            Display::Grid |
+            Display::WebkitBox => *self,
+
+            // Everything else becomes block.
+            _ => Display::Block,
+        }
+
+    }
+
+    /// Convert this display into an inline-outside display.
+    ///
+    /// Ideally it should implement spec: https://drafts.csswg.org/css-display/#inlinify
+    /// but the spec isn't stable enough, so we copy what Gecko does for now.
+    #[cfg(feature = "gecko")]
+    pub fn inlinify(&self) -> Self {
+        match *self {
+            Display::Block |
+            Display::FlowRoot => Display::InlineBlock,
+            Display::Table => Display::InlineTable,
+            Display::Flex => Display::InlineFlex,
+            Display::Grid => Display::InlineGrid,
+            Display::MozBox => Display::MozInlineBox,
+            Display::MozStack => Display::MozInlineStack,
+            Display::WebkitBox => Display::WebkitInlineBox,
+            other => other,
+        }
+    }
+
+    #[cfg(feature = "servo")]
+    #[inline]
+    /// Custom cascade for the `display` property in servo
+    pub fn cascade_property_custom(
+        _declaration: &PropertyDeclaration,
+        context: &mut Context
+    ) {
+        longhands::_servo_display_for_hypothetical_box::derive_from_display(context);
+        longhands::_servo_text_decorations_in_effect::derive_from_display(context);
+    }
+}
+
+impl Parse for Display {
```

Comment:
You should be able to `derive` this, shouldn't you?

Reply:
Ooooooh, yeah, I think I should try to use `derive` for it.
However, I have only one concern to make it use `derive`.
For `Display::Flex` and `Display::InlineFlex`, they are originally matches with `-webkit-` prefix.
If we try to derive `Parse`, I think it will be matched without the prefix?

Reply:
Just update the PR with `derive` `Parse`!
I'd like to know if there's any test error with it :)

Reply:
Oh, that will definitely cause tests to fail. I opened  #19665 with a solution for that.

Reply:
WOW! Great solution! Let me try to fix it!

Reply:
By the way, the [CI build](https://travis-ci.org/servo/servo/builds/323504005) in the patch without `#[parse(alias = "-webkit-flex")]` still succeeded.
I guess it didn't be tested. Shall we have another PR to add test for it?

Reply:
Sure, the CI build succeeds, but that's only a portion of the tests we run. To run all the tests you should push to Mozilla's try server (https://wiki.mozilla.org/ReleaseEngineering/TryServer), which I think you don't have commit access to because it's a Gecko thingie.

For PRs that just move code around I don't use to push them if I'm confident the changes are correct (we run the tests when the patch lands anyway), but for bigger changes I push and get a link like: https://treeherder.mozilla.org/#/jobs?repo=try&revision=9f764414e0f2341c4494164b8341ab9d901e7524

Reply:
Thanks for the information.